File: app/src/main/java/au/com/codeka/weather/providers/wunderground/WundergroundResponse.java
package au.com.codeka.weather.providers.wunderground;

import androidx.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

/**
 * POJO which we use Gson to decode the weather underground response into.
 */
@SuppressWarnings("unused")
public class WundergroundResponse {
  @SerializedName("current_observation") CurrentObservation currentObservation;
  Forecast forecast;
  @SerializedName("hourly_forecast") ArrayList<HourlyForecast> hourlyForecast;

  public static class CurrentObservation {
    ObservationLocation observationLocation;
    String stationId;
    @SerializedName("observation_epoch") String observationTime; // seconds since epoch
    String weather; // e.g. "Light Rain", "Sunny" etc.
    @SerializedName("temp_c") @Nullable Double temp; // temperature in degrees celsius
    String relativeHumidity;
    String windDir; // e.g. N, NE, etc.
    @Nullable Double windKph;
    @Nullable Double windGustKph;
    @SerializedName("feelslike_c") @Nullable Double feelsLike; // "feels like" temperaure in degrees celcius
    @SerializedName("precip_1hr_metric") @Nullable Double precipitationLastHour; // in mm
    @SerializedName("precip_today_metric") @Nullable Double precipitationToday; // in mm
    String icon;
  }

  public static class ObservationLocation {
    String full;
    String city;
    String state;
    String country;
    String lat;
    String lng;
    String elevation;
  }

  public static class Forecast {
    TextForecast txtForecast;
    @SerializedName("simpleforecast") SimpleForecast simpleForecast;
  }

  public static class TextForecast {
    @SerializedName("forecastday") ArrayList<TextForecastDay> days;
  }

  public static class TextForecastDay {
    int period;
    String icon;
    String title;
    @SerializedName("fcttext_metric") String forecastText;
  }

  public static class SimpleForecast {
    @SerializedName("forecastday") ArrayList<SimpleForecastDay> days;
  }

  public static class SimpleForecastDay {
    int period;
    SimpleForecastTemp high;
    SimpleForecastTemp low;
    String conditions;
    String icon;
  }

  public static class SimpleForecastTemp {
    String fahrenheit;
    String celsius;
  }

  public static class HourlyForecast {
    @SerializedName("FCTTIME") HourlyForecastTime time;
    HourlyForecastValue temp;
    String icon;
    @SerializedName("wc") String shortDescription;
    HourlyForecastValue qpf;
  }

  public static class HourlyForecastValue {
    String english; // who is naming these, seriously??
    String metric;
  }

  public static class HourlyForecastTime {
    public String hour;
    public String epoch;
  }
}

File: app/src/main/java/au/com/codeka/weather/providers/wunderground/WundergroundResponse.kt
package au.com.codeka.weather.providers.wunderground

import com.google.gson.annotations.SerializedName
import java.util.*

/**
 * POJO which we use Gson to decode the weather underground response into.
 */
class WundergroundResponse {
  @SerializedName("current_observation")
  var currentObservation: CurrentObservation? = null
  var forecast: Forecast? = null

  @SerializedName("hourly_forecast")
  var hourlyForecast: ArrayList<HourlyForecast>? = null

  class CurrentObservation {
    var observationLocation: ObservationLocation? = null
    var stationId: String? = null

    @SerializedName("observation_epoch")
    var observationTime // seconds since epoch
        : String? = null
    var weather // e.g. "Light Rain", "Sunny" etc.
        : String? = null

    @SerializedName("temp_c")
    var temp // temperature in degrees celsius
        : Double? = null
    var relativeHumidity: String? = null
    var windDir // e.g. N, NE, etc.
        : String? = null
    var windKph: Double? = null
    var windGustKph: Double? = null

    @SerializedName("feelslike_c")
    var feelsLike // "feels like" temperaure in degrees celcius
        : Double? = null

    @SerializedName("precip_1hr_metric")
    var precipitationLastHour // in mm
        : Double? = null

    @SerializedName("precip_today_metric")
    var precipitationToday // in mm
        : Double? = null
    var icon: String? = null
  }

  class ObservationLocation {
    var full: String? = null
    var city: String? = null
    var state: String? = null
    var country: String? = null
    var lat: String? = null
    var lng: String? = null
    var elevation: String? = null
  }

  class Forecast {
    var txtForecast: TextForecast? = null

    @SerializedName("simpleforecast")
    var simpleForecast: SimpleForecast? = null
  }

  class TextForecast {
    @SerializedName("forecastday")
    var days: ArrayList<TextForecastDay>? = null
  }

  class TextForecastDay {
    var period = 0
    var icon: String? = null
    var title: String? = null

    @SerializedName("fcttext_metric")
    var forecastText: String? = null
  }

  class SimpleForecast {
    @SerializedName("forecastday")
    var days: ArrayList<SimpleForecastDay>? = null
  }

  class SimpleForecastDay {
    var period = 0
    var high: SimpleForecastTemp? = null
    var low: SimpleForecastTemp? = null
    var conditions: String? = null
    var icon: String? = null
  }

  class SimpleForecastTemp {
    var fahrenheit: String? = null
    var celsius: String? = null
  }

  class HourlyForecast {
    @SerializedName("FCTTIME")
    var time: HourlyForecastTime? = null
    var temp: HourlyForecastValue? = null
    var icon: String? = null

    @SerializedName("wc")
    var shortDescription: String? = null
    var qpf: HourlyForecastValue? = null
  }

  class HourlyForecastValue {
    var english // who is naming these, seriously??
        : String? = null
    var metric: String? = null
  }

  class HourlyForecastTime {
    var hour: String? = null
    var epoch: String? = null
  }
}